### DISCUSSION TITLE
derive() with get(store, 'key')

### DISCUSSION THREAD

**Discussion by sarimarton** [pmndrs/valtio#714] [Ideas]
I think it would be a nice complimentary addition to the `get()` function in derive, if we could specify a key so it turns into a key subscription.

So instead of

```tsx
derive(
  {
    tripled: (get) => get(state).count * 3,
  },
  { proxy: state }
)
```

It would be:

```tsx
derive(
  {
    tripled: (get) => get(state, 'count') * 3,
  },
  { proxy: state }
)
```

The benefit of the latter one is that it doesn't have unnecessary recalculation. The possible keys could be typed with typescript. I would also love to have nested keys there, e.g.:

```tsx
derive(
  {
    tripled: (get) => get(state, 'some.nested.key') * 3,
  },
  { proxy: state }
)
```

Which is totally doable with typescript. This pattern focuses on the leaf values, which are typically the primary candidates for dependencies. It matches with Valtio's snapshot tracking logic as well, which also focuses on leaf values.

If there's positive feedback, i would gladly work on it. Unfortunately `derive()` is too complex to make this in userland without a huge amount of copy-paste, so I'm proposing it here.

**Comment by dai-shi** [comment]
> Unfortunately `derive()` is too complex to make this in userland without a huge amount of copy-paste

Well, that's the thing. derive is too complex that we can no longer reasonably maintain. So, I can't accept adding features. I'm hoping to rather simplify it. So, for now, you should implement your idea as a third-party package and list in https://github.com/pmndrs/valtio/blob/main/docs/resources/libraries.mdx.

btw, you might be interested in https://github.com/dai-shi/use-valtio, which does something for nested keys. I found that TypeScript typing was very tricky.
